File: Easy/Find if Path Exists in Graph/find-if-path-exists-in-graph.cpp
class Solution {
public:
    bool validPath(int n, vector<vector<int>>& edges, int src, int dst) {
        if(src == dst) return true;

        vector<unordered_set<int>> graph(n);
        for(auto edge : edges)
        {
            graph[edge[0]].insert(edge[1]);
            graph[edge[1]].insert(edge[0]);
        }

        set<int> seen;
        queue<int> nodes;
        nodes.push(src);

        while(!nodes.empty())
        {
            queue<int> tmp;
            while(!nodes.empty())
            {
                int node = nodes.front();
                nodes.pop();
                for(auto x : graph[node])
                {
                    if(seen.find(x) == seen.end())
                    {
                        if(x == dst) return true;
                        seen.insert(x);
                        tmp.push(x);
                    }
                }
            }
            nodes = tmp;
        }

        return false;
    }
};

File: Easy/Find if Path Exists in Graph/find-if-path-exists-in-graph.py
class Solution:
    def validPath(self, n: int, edges: List[List[int]], src: int, dst: int) -> bool:
        
        if src == dst: return True
        
        graph = defaultdict(list)
        for edge in edges:
            graph[edge[0]].append(edge[1])
            graph[edge[1]].append(edge[0])

        seen = set()
        queue = deque([src])

        while queue:
            tmp = deque([])
            while queue:
                node = queue.popleft()
                for e in graph[node]:
                    if e not in seen:
                        if e == dst: return True
                        seen.add(e)
                        tmp.append(e)
            queue = tmp

        return False
